### DISCUSSION TITLE
Some Atari 7800 carts dump fine with hw3, not with hw5

### DISCUSSION THREAD

**Discussion by PsyK0p4T** [sanni/cartreader#878] [Hardware]
Hey people,

Creating this thread to share some behaviors I had here.
I've got two kits here, one hw3 and a hw5. None of them were built by me, so won't be able to precisely discuss about the components used in both. Just never had any particular issues with them until now, and I've dumped a large quantity of games/systems.

My issue came with Atari 7800. I'm using the A7800 to SNES adapter.
Carts for this system come in different capacities.
Smallest ones (Donkey Kong Junior, Super Huey...) are 48KB, and can be dumped fine on hw3 AND hw5.
Higher ones, like 128KB (Commando, Ace of Aces...) recquire bank switching. For these ones, only my hw3 managed to dump them fine. The hw5 seems to not bank-switch properly.

With these results, the code seems out of cause, as the SNES slot and A7800 adapter.
So the problem would be more coming from the reader itself, hardware side ?
Could the voltage be the problem here ? Is the hw3 known to deliver a bit more than hw5 in general ?
I don't have big equipment like an oscilloscope here. But if there are some tests I can do on the SNES slot, with a  voltmeter for example, just tell me. Maybe old systems/chips are more greedy and ask for more power to run properly. Any of you ever encoutered this ?

**Comment by sanni** [comment]
On HW3 the SNES slot's CPU CLOCK pin can be connected to either the Adafruit Clock Generator or the Atmega2560's PH1 pin by the CLK1 switch. 
HW5 behaves like a HW3 with the two CLK switches constantly set to the ON position (=connected to clock generator).
With the switches set like this your HW3 should also fail to dump Atari 7800 cartridges that need a bank switch.
The reason being that the [bankSwitch_7800() function](https://github.com/sanni/cartreader/blob/master/Cart_Reader/7800.ino#L249) uses the PH1 pin.
This means the code and the 7800 adapter PCB need to be updated to work on HW5.

**Comment by PsyK0p4T** [comment]
Thanks for your feedback sanni !
Unfortunately, I won't be able to test your point for the moment. I had borrowed these carts from a friend, and had to return them. I'll try to buy some for future uses.

**Comment by sanni** [comment]
It could also be possible to leave the adapter PCB as is and try to use the Clock Generator in the [bankSwitch_7800() function](https://github.com/sanni/cartreader/blob/master/Cart_Reader/7800.ino#L249). Possibly the preferred way.
I think we had something like this before with some other system and there using the Clock Generator instead of PH1 did indeed work.

**Comment by smesgr9000** [comment]
thanks for the hint @PsyK0p4T . I will try to investigate further.

**Comment by smesgr9000** [comment]
I got the following games working with this and the dump is identical to the Atari Max:
Supergames:
Crack'ed
Mat Mania Challenge
Motor Psycho
Scrapyard Dog

Normal:
Ballblazer
Donkey Kong


[PATCH.txt](https://github.com/sanni/cartreader/files/14156755/PATCH.txt)

**Reply by smesgr9000** [reply]
@PsyK0p4T   do you have a chance to test the above patch?

**Reply by PsyK0p4T** [reply]
Hey,
Unfortunately, not yet. I don't have the cartridges in hands at the moment, and must borrow them back from a friend. I hope to get them soon... 
Can you maybe post the full 7800.ino file with your updates already merged in ? It will be faster ;)

**Reply by smesgr9000** [reply]
Hey 
sure, but be aware if other stuff changes the file may not be compatible anymore. You have to rename it back to ino due to github doesn't allow ino file endings.
[7800.txt](https://github.com/sanni/cartreader/files/14323756/7800.txt)

**Reply by PsyK0p4T** [reply]
Good news here !
I had the time to get back some cartridges and test your code. It's working perfectly ! With previously supported games, and also the ones which had timing problems. Thanks for this contribution, it will be merged soon :)

Tested and working : Crack'ed, Ikari Warriors, Donkey Kong Junior, Super Huey, Commando, Ace of aces.

**Reply by smesgr9000** [reply]
nice thanks

**Comment by KCgamester** [comment]
I'm still having very little success with dumping my 7800 carts.. I have hardware version 5 with RTC,  Clock Generator, firmware V15.4, and Vselect. Roughly 90% of my 7800 games will not dump properly.  Any advice would be appreciated.

**Reply by cmcm711** [reply]
Make sure you are cleaning the PCB edge on your cartridges thoroughly to the point where the cotton swabs come back with no dirt on them.

**Reply by KCgamester** [reply]
All my carts are clean. The brand new 7800+ games wont even dump.

**Reply by cmcm711** [reply]
I wonder if that is the issue? I’m not sure if “new” Atari 7800+ are dumpable with the OSCR.

Has anyone else tried this?

**Reply by NML32** [reply]
> I'm still having very little success with dumping my 7800 carts.. I have hardware version 5 with RTC, Clock Generator, firmware V15.4, and Vselect. Roughly 90% of my 7800 games will not dump properly. Any advice would be appreciated.

I'm also having issues dumping Atari 7800 cartridges. I have the same version dumper. I have v1 of the 7800 cart adapter. I ordered version 2 of the 7800 cart adapter from eBay. I'm hoping that will make a difference. A.R.T.I. is listed in the A7800 dump list, but I can't get a good dump. 

**Update:** I was able to dump a working copy of A.R.T.I.  After I set the header I was able to load and run on my MiSTer FPGA. 

7800 A78 Header Info   : atari.a78

    embedded game name : A.R.T.I
    rom size           : 262144
    cart format        : SuperGame ram@4000
    controllers        : 7800joy1 7800joy2
    save peripheral    : none
    xm/xboard          : not enabled
    tv format          : NTSC

**Reply by smesgr9000** [reply]
OSSC with RTC is sadly still very unreliable for 7800 which require Bank switching. I have tested several methods to get rid of the cloak all together. ATM none is really nice solutions. I hope to invest some more time into this over the holidays. If you can use ossc without RTC. Btw yes the 7800+ games dump fine - i don' have tested all rereleases but those i got my hands work for me - of course they are also affected by the RTC issue.

**Reply by bootsector** [reply]
> > I'm still having very little success with dumping my 7800 carts.. I have hardware version 5 with RTC, Clock Generator, firmware V15.4, and Vselect. Roughly 90% of my 7800 games will not dump properly. Any advice would be appreciated.
> 
> I'm also having issues dumping Atari 7800 cartridges. I have the same version dumper. I have v1 of the 7800 cart adapter. I ordered version 2 of the 7800 cart adapter from eBay. I'm hoping that will make a difference. A.R.T.I. is listed in the A7800 dump list, but I can't get a good dump.
> 
> **Update:** I was able to dump a working copy of A.R.T.I. After I set the header I was able to load and run on my MiSTer FPGA.
> 
> 7800 A78 Header Info : atari.a78
> 
> ```
> embedded game name : A.R.T.I
> rom size           : 262144
> cart format        : SuperGame ram@4000
> controllers        : 7800joy1 7800joy2
> save peripheral    : none
> xm/xboard          : not enabled
> tv format          : NTSC
> ```

Did you remove the clock generator from your OSCR or did you just disable it in Config.h?

**Reply by NML32** [reply]
I didn't remove the clock generator or disable it in the config after I dumped A.R.T.I.  I used 7800header.exe to set the header to SuperGame ram@4000. Then I could load the game on my MiSTer FPGA.  
These are the settings I'm using.  
<img width="932" height="1141" alt="Screenshot 2025-12-25 150306" src="https://github.com/user-attachments/assets/4f609dcc-1553-48dd-809b-a93b8c64fbd0" />

**Reply by bootsector** [reply]
That's interesting. Are you able to get the same ROM contents dump (hash verified) three times in a row? From looking at the current code, the clockgen will interfere with PH1 and getting a good dump would be just luck?

**Reply by NML32** [reply]
I dumped 4 times and got the same hash CRC32 ebe905a1
I didn't build my OSCR. I checked with the person who built it to verify that it has a clockgen module.

**Reply by bootsector** [reply]
Thanks! Just figured out that not all games use the bank-switching mechanism that uses the clockgen, so for those games they should dump just fine.

**Comment by bootsector** [comment]
Waiting for a couple of new released Atari 7800 games to arrive. Got myself a v2 Atari 7800 adapter. What should I do in order to make the dumping process more reliable while I wait? Would disabling CLOCKGEN and RTC in Config.h help?

**Comment by bootsector** [comment]
Taking a look at the code here, I'm wondering why is RTC affecting the 7800 functionality at all? I see references in the code to the CLOCKGEN but not RTC.

**Comment by bootsector** [comment]
Just opened a PR that might help with this: https://github.com/sanni/cartreader/pull/1150

**Reply by smesgr9000** [reply]
yes sorry the Clockgen not the RTC is the issue. I was distracted writing the post. Your PR will likely not helping. If I recall correctly the Clockgen pulls the PIN either low or high if disabled. Thus the Ardurino still can't control the PIN. My Workaround was to instead pull the PIN itself use the Clockgen to move into a constant high or low state. But this isn't really good coding standard. I was thinking on disabling the clockgen in high impedance state and than use the non-clockgen code. But for some reason this didn't work. But I haven't had enough time yet to figure out whats going on, maybe just a silly bug.

**Reply by bootsector** [reply]
Oh, so we need to physically remove the clock gen from the OSCR? Yep, my PR won’t work for sure then.

**Comment by bootsector** [comment]
@smesgr9000 thanks for your insights. I've updated my PR with the code that should put the clock gen in HI-Z mode if disabled. This is fully NOT tested yet though as I'm still waiting for my carts to arrive.

**Comment by bootsector** [comment]
Stupid question: do we need a clock generator for dumping any 7800 game at all?

**Comment by KCgamester** [comment]
For the record, disabling the RTC or physically removing the battery does not work.

**Reply by bootsector** [reply]
Yep. The issue is not related to the RTC but the clock generator

**Comment by bootsector** [comment]
@KCgamester @cmcm711 just in case you (or anyone else having a HW5 with a clock generator) want to give my PR a try, just download the current release zip package and update the files from the zip attached to this post before configuring/building/flashing. 

Please make sure your config has ENABLE_CLOCKGEN enabled.

Test reports are hugely appreciated. Thanks!
[PR.zip](https://github.com/user-attachments/files/24339099/PR.zip)

**Comment by sanni** [comment]
Try this [7800.zip](https://github.com/user-attachments/files/24340090/7800.zip)

It mimics how HW3 with "CLK1 switch set to off" does bank switching but uses the clkgen: clk pin is set high by default and only pulses a couple of times during bank switching. 

Top HW5, bottom HW3:
![Untitled](https://github.com/user-attachments/assets/be5dd8db-997d-44b9-a3ef-eaf3ee91bc81)

I set the clk to 10 kHz to get 9 pulses similar to the 8 on HW3.

**Reply by bootsector** [reply]
Do you still get those pulses if you set the clockgen output as HI-Z? Isn't dangerous for the clockgen to be driving the clock pin HIGH while the Arduino drives the pin connected to it LOW?

**Reply by sanni** [reply]
On HW3 there is a physical switch that selects between either the clock generator's CLK1 output or the Atmega's PH1 pin to drive the clk line that is connected to the 7800 adapter. 

HW5 does not have this switch so only the clock generator's CLK1 output is connected to the 7800 adapter. The Atmega's PH1 pin is not connected at all to the SNES slot.

We could remake the 7800 adapter and choose a pin that both HW3 and HW5 have exposed on the SNES slot or we can fix it in software.

**Reply by bootsector** [reply]
> On HW3 there is a physical switch that selects between either the clock generator's CLK1 output or the Atmega's PH1 pin to drive the clk line that is connected to the 7800 adapter.
> 
> HW5 does not have this switch so only the clock generator's CLK1 output is connected to the 7800 adapter. The Atmega's PH1 pin is not connected at all to the SNES slot.
> 
> We could remake the 7800 adapter and choose a pin that both HW3 and HW5 have exposed on the SNES slot or we can fix it in software.

Thanks! I completely misunderstood the pin connections. I understand now what your code is doing inside the bankswitch function. And I now wonder how people can still bankswitch the carts with the current code? haha

I will go ahead and remove my PR!

**Reply by KCgamester** [reply]
@sanni I just tried your 7800.ino file.  No luck.  I'm stll getting bad dumps.

**Reply by bootsector** [reply]
@KCgamester what version is your 7800 adapter? Also, can  you give me the name of a game you own that it's failing to dump, preferably one that is available for purchase on eBay or whatever, so I can purchase it and test it on my side as well?

**Reply by KCgamester** [reply]
> @KCgamester what version is your 7800 adapter? Also, can you give me the name of a game you own that it's failing to dump, preferably one that is available for purchase on eBay or whatever, so I can purchase it and test it on my side as well?

I have two 7800 adaters,. Version 01 and 02. The following games will not dump properly (black screen or corrupt graphics with loud noises on the ProSytem emulator and Mister FPGA). **Using a 7800 Game Drive on original hardware, EVERY 7800 GAME I've dumped has an "Invalid File" error and WILL NOT LOAD.** Version 02 adapter was used today for testing.

1. Pole Position II
2. Dig Dug
3. Fight Night
4. Impossible Mission (wont play past title screen)
5. One On One
6. Food Fight
7. Tower Toppler
8. Galaga
9. Alien Brigade (only works on Prosystem Emulator)
10. Robotron
11. Tiger Heli  7800+
12. Pac Man Double Feature 7800+
13. Bentley Bear's Crystal Quest 7800+
14. Bentley Bear's Crystal Quest Atari Age 
15. Commando (Mister FPGA - the pokey sound is missing and on ProSytem emulator the title screen has graphic errors).
16. Ballblazer (Mister FPGA - the pokey sound is missing. Works on ProSystem Emulator).

The following games were tested and working on **Mister FPGA and Pro System emulator ONLY**.

1. Karateka
2. Joust
3. Desert Falcon
4. Xevious
5. Hat Trick
6. Centedepe
7. Midnight Mutants
8. Choplifter

**Reply by bootsector** [reply]
Thanks! I have two games from your non-working list coming in in the next few days. I will test and report back. 

Are the dumps consistent (same hash) if you dump any of them 3 times? Wondering if the emulators need a header in order to properly detect the mapper scheme used by the game?

Finally, just to make sure your clockgen is good, are you able to dump a SA-1 SNES game correctly? If you need a cheap game to test,  the Japanese version of Super Mario RPG can be found for $10-ish on eBay.

**Reply by KCgamester** [reply]
The dumps do not produce consistent crc codes.  Very rarely will the same crc appear for the same game dumped multiple times.

**Reply by smesgr9000** [reply]
1. Not all of your mentioned games are bank switching games. Thus the described issue here does not affect those games. Everything with "Standard" does not do bank switching and does not do anything with CLK-Signal. (some of your listed are: Dig Dug, One-on-One, Robotron, Ballblazer, PacMan) - and other way around Midnight Mutants is a bank switching game
2. V1 and V2 of the 7800 adapter does not make any difference for this. The difference is connection of an additional PIN only used by "Bankset switching" games introduces with Petscii Robot. Only a few homebrew games use this mapper.
3. CLKgen does not prevent dumping bank switching games. But it is possible to produce a race condition between clk gen and the adruino. Thus dumping is possible but you may have to retry several times.
4. Missing Pokey sound has nothing to do with dumping. This points to wrong header information

Thus it is very likely you have another issue not related. Check if you have:
- 5V
- the cartridge in the right orientation
- properly aligned PCB. I'm using a Wingtat connector and there it is possible to shift the edge of the cartridge left and right losing contact with the connector pins.
- cleaned your edge connectors
- you have selected the right mapper

**Reply by smesgr9000** [reply]
@sanni :
> HW5 does not have this switch so only the clock generator's CLK1 output is connected to the 7800 adapter. The Atmega's PH1 pin is not connected at all to the SNES slot.

ah this explains why my workaround did not work. Thats bad. So the only software solution I see is my hack:
```
#ifdef ENABLE_CLOCKGEN
set_clock_disable(SI5351_CLK1, SI5351_CLK_DISABLE_HIGH);
#else
PORTH |= (1 << 1);  // CLK(PH1) HIGH
#endif

#ifdef ENABLE_CLOCKGEN
set_clock_disable(SI5351_CLK1, SI5351_CLK_DISABLE_LOW);
#else
PORTH &= ~(1 << 1); // CLK(PH1) LOW
#endif
```

its bad because several adapters by @sakman55 uses this pin.

**Reply by bootsector** [reply]
@smesgr9000 there’s a potential good fix here by @sanni : https://github.com/sanni/cartreader/discussions/878#discussioncomment-15342282

It would be great if people could test it with carts that need bank switching so that could finally be merged

**Reply by smesgr9000** [reply]
Don't think thats a fix, more a workaround.

**Reply by NML32** [reply]
Pole Position II 7800 cart fails for me as well when I try to dump from my v5 OSCR.

**Reply by bootsector** [reply]
@NML32 Pole Position II is a non-bankswitching game. Did you test your game on the OG console?

**Reply by NML32** [reply]
Yes, the cart works on my OG 7800 console and Atari 2600+ Pac-Man edition.

**Reply by bootsector** [reply]
@NML32 thanks for confirming! I will purchase that cart on eBay so I can test myself as well!

**Reply by bootsector** [reply]
> Try this [7800.zip](https://github.com/user-attachments/files/24340090/7800.zip)
> 
> It mimics how HW3 with "CLK1 switch set to off" does bank switching but uses the clkgen: clk pin is set high by default and only pulses a couple of times during bank switching.
> 
> Top HW5, bottom HW3: ![Untitled](https://private-user-images.githubusercontent.com/19169494/530230715-be5dd8db-997d-44b9-a3ef-eaf3ee91bc81.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcwNDE4OTgsIm5iZiI6MTc2NzA0MTU5OCwicGF0aCI6Ii8xOTE2OTQ5NC81MzAyMzA3MTUtYmU1ZGQ4ZGItOTk3ZC00NGI5LWEzZWYtZWFmM2VlOTFiYzgxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjI5VDIwNTMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0YmU5ZmY4OTZiNTE4N2NhMzY3MGY2MzA1NDg3MjhhNGMxOGEyOWJhOGQ1Njc2MjBmYTA4MjNlMDFmZTgwNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tB5aLw2rQ0dDf1yOJ00NpM5rwlyOAyCvJMxMsX5Bqo8)
> 
> I set the clk to 10 kHz to get 9 pulses similar to the 8 on HW3.

This fix worked great with both Tiger-Heli 7800 and Pac-Man Double Feature 7800. 

These are the header options I'm setting with 7800header so I get Tiger-Heli 7800 working on MiSter FPGA:

supergame rom@4000 pokey@450

**Reply by NML32** [reply]
@bootsector With your header info and updated 7800.ino I was able to dump and play Tiger-Heli. I do see some garbage on the high-score table, and I never get prompted to add my initials at the end of the game. I don't know if there are any other graphic glitches since I'm not very good at the game and don't advance very far. 
What Mapper setting is used for the Pacman double feature cart?

<img width="1538" height="1080" alt="20251229_170501-Tiger-Heli" src="https://github.com/user-attachments/assets/a7d0b2e0-a514-458f-b0f4-5a1b4f20f8f5" />

**Reply by bootsector** [reply]
This is the ROM info I get for Tiger-Heli before applying a header to it:

> File size: 147456 bytes
> CRC-32   : 56ACF5B7
> MD5      : 5D9AC066E685432825D968F166B042EC
> SHA-1    : 7F96BA3E9E497BD220098146DEDBE8F1A3C250F8
> SHA-256  : ACC64ED224E026E271C26FD7B734DD3F642AB951516F2B7972C622483D475544

Do you get the same? I will test further on the MiSTer when I get a chance.

**Reply by bootsector** [reply]
For Pac Man I just selected Pac-Man from the games list

**Reply by NML32** [reply]
Yes, my Tiger-Heli dump before the header applied. 
MD5: 5d9ac066e685432825d968f166b042ec
SHA1: 7f96ba3e9e497bd220098146dedbe8f1a3c250f8
SHA 256: acc64ed224e026e271c26fd7b734dd3f642ab951516f2b7972c622483d475544

**Reply by bootsector** [reply]
@NML32 the scores from your screenshot look crazy high. I can't reproduce the issue here. It allows me to put my initials just fine and I see no graphical glitches.

This is what I get on the Atari7800 (24.04.23) MiSTer core:

![IMG_5409](https://github.com/user-attachments/assets/20da8054-6df0-4683-9b67-a065af4f0dd1)

This is the header info according to 7800header version 0.21 Aug 10 2023 13:57:54:

<img width="370" height="107" alt="image" src="https://github.com/user-attachments/assets/7adceb08-7cb2-48c6-b81d-89722ce62633" />

**Reply by NML32** [reply]
@bootsector Thank you for testing. I found the issue. If I have Save Key enabled, I have the high-score issue. 
Under Atari 7800 settings 
Peripherals > Port2 Input: SaveKey  
I set it back to Auto, and all is good. 
Weird.

**Reply by KCgamester** [reply]
Update!

1. I've discovered that my cart reader (Aliexpress) has a small gap in the SNES port, making the connection unrealiable.  I have to ensure that the 7800 adapter is adjusted furthest to the left in order to dump properly.
2. All my carts with no mapper now work in both the Mister FPGA and the 7800 Game Drive, with the exception of Pole Position II. I was able to get Ballblazer's pokey to work by entering pokey@4000 in the header.
3. By using the header info provided by @bootsector and the .ino file provided by @Sanni , Bently Bear's Crystal Quest and Tiger Heli both now work in Mister FPGA and the 7800 Game Drive.

Thanks!!!

**Reply by bootsector** [reply]
Another report about Pole Position II not dumping correctly! That's very weird! Looking forward to my cart to arrive from eBay so I can take a look at it myself.

**Comment by bootsector** [comment]
@sanni wondering if the additional 1 pulse would affect anything? For functionality parity and peace of mind, would it be possible to slow down the frequency of the clockgen a little bit aiming for 8 pulses in that interval?

**Comment by PW5190** [comment]
I'm considering on getting a 7800 adapter to dump new cartridges produced by Atari. I'm thinking of getting two from Atari directly: _Mr. Run and Jump 2600_ and _Pac-Man Double Feature 7800_.

**Comment by bootsector** [comment]
Had no issues dumping Pole Position II, although we do have two reports here from people not being able to dump it:

<img width="3024" height="4032" alt="image" src="https://github.com/user-attachments/assets/99c946fe-3929-4290-919e-d279ff391b44" />

<img width="474" height="278" alt="image" src="https://github.com/user-attachments/assets/9a629eef-cd71-47cd-90b2-4d5fb409800e" />

<img width="786" height="72" alt="image" src="https://github.com/user-attachments/assets/d11c255b-c891-4230-bf0a-44d242c0b866" />

<img width="2170" height="226" alt="image" src="https://github.com/user-attachments/assets/14e34dc4-bebd-4d39-9ed2-7eeb0883d7bd" />

Wondering what could be causing issues to some people as this game doesn't use bank-switching at all. 🤔

**Reply by KCgamester** [reply]
![20260108_042704.jpg](https://github.com/user-attachments/assets/eecdc3bd-0550-495f-ba48-ccb2b79e89ce)
![20260108_042407.jpg](https://github.com/user-attachments/assets/8aa06acf-af70-4d07-ba8c-d987fe84b907)

I was able to dump Pole Position II using adapter v 0.1, but not on v 0.2. Weird.

**Reply by bootsector** [reply]
Pretty sure this is something physical related to the connector and the cartridge and not software/firmware related.

**Reply by KCgamester** [reply]
Agreed.
